### ISSUE TITLE
No module named Hyperspace on AWS EMR

### ISSUE THREAD

**Issue by mmchougule** [microsoft/hyperspace#297]
### Describe the issue
Great project, I am trying to use this in pyspark on AWS EMR. I have added `spark.jars.packages: "com.microsoft.hyperspace:hyperspace-core_2.11:0.1.0"` in my spark config and is used when creating a SparkSession with spark master on `yarn`.
When I do `from hyperspace import Hyperspace` I am getting an error `No module named hyperspace`. I do see that it has downloaded the hyperspace jar but for some reason, my code isn't recognizing this library in python. Any ideas what I could be doing wrong?

### To Reproduce
1. Add hyperspace jar when creating a spark session in jupyter notebook or in the project code
2. Add `from hyperspace import Hyperspace`, you will see the error.

### Expected behavior
I should be able to import Hyperspace so I can use it.

### Environment 
AWS EMR, centos VM

**Comment by apoorvedave1** [comment]
Hi @mmchougule , could you please try the following; 

C:\path\to\hyperspace\jar> pyspark --jars hyperspace-core_2.11-<version>.jar --files hyperspace-core_2.11-<version>.jar
…
Spark session start logs
…
>>> from hyperspace import *
>>> hyperspace = Hyperspace(spark)
>>>

**Comment by mmchougule** [comment]
Hi @apoorvedave1 thanks for getting back to me. This works for me in the command line. I see the issue  I run it in my code or in a jupyter notebook and try to create a SparkSession with spark.jars as the config. Am I doing it wrong? I also tried passing the path of the jar file on the VM. It seems like it did download the jar correctly just not enabling the `hyperspace` package for some reason.
```
spark = (
    SparkSession.builder.master("yarn")
    .config("spark.jars", "com.microsoft.hyperspace:hyperspace-core_2.11:0.3.0")
    .appName("myapp")
    .getOrCreate()
)
 
```

**Comment by apoorvedave1** [comment]
thanks @mmchougule , I am not really sure how to move forward in this case. Could you try replacing hyperspace jar with any other external jar in your environment? 
I am not familiar with setting up jupyter with external jars and just wanted to rule out the possibility that it's a problem with hyperspace jar or just its availability in the environment.

**Comment by apoorvedave1** [comment]
We would also love your contribution if you manage to figure out how to setup Jupyter with hyperspace. Please feel free to add instructions to do it if it works for you!

**Comment by mmchougule** [comment]
Sounds good. I will debug and try a few more things. I can contribute back once I figure it out. Thank you!

**Comment by sezruby** [comment]
@mmchougule Could you try this in a jupyter notebook?
```
%%configure -f
{ "conf": {"spark.jars.packages": "com.microsoft.hyperspace:hyperspace-core_2.11:0.3.0" }}
```

**Comment by mmchougule** [comment]
Thanks @sezruby It is showing me the Cell magic `%%configure` not found. Trying to find how to enable that so I can see if above works.

**Comment by mmchougule** [comment]
Just an update:
1. Last night, I was able to use it in a notebook. I created indexes but the explain was still showing me sortMergeJoin and exchange shuffle partitions. This is for a join query. e.g. `joined = df1.join(df2, [id_col, date_col, "another_id"], [])`. hs.explain() did not show much either. This did seem to work when I just do a filter by id_col and do a select.

2. Today, I was seeing weird errors on running the same code of creating indexes. I was seeing the same jackson error when I do hs.indexes().show() and was not able to deleteIndex either. I didn't capture that stacktrace, I did delete hdfs -dfs -rm -r the `/spark/warehouse/indexes/index` directory so I start fresh but now creating index shows me the following error. Any ideas?

```An error occurred while calling o532.createIndex.
: com.fasterxml.jackson.databind.JsonMappingException: Illegal unquoted character ((CTRL-CHAR, code 0)): has to be escaped using backslash to be included in name
 at [Source: java.io.StringReader@51ef8888; line: 980, column: 23] (through reference chain: com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["content"]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["root"]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["subDirs"]->com.fasterxml.jackson.module.scala.deser.BuilderWrapper[0]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["subDirs"]->com.fasterxml.jackson.module.scala.deser.BuilderWrapper[0]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["subDirs"]->com.fasterxml.jackson.module.scala.deser.BuilderWrapper[0]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["subDirs"]->com.fasterxml.jackson.module.scala.deser.BuilderWrapper[0]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["subDirs"]->com.fasterxml.jackson.module.scala.deser.BuilderWrapper[0]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["subDirs"]->com.fasterxml.jackson.module.scala.deser.BuilderWrapper[0]->com.fasterxml.jackson.module.scala.deser.MapBuilderWrapper["files"]->com.fasterxml.jackson.module.scala.deser.BuilderWrapper[0])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:210)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:189)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:266)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:217)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:59)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapArray(UntypedObjectDeserializerModule.scala:39)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:233)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:217)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:59)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapArray(UntypedObjectDeserializerModule.scala:39)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:233)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:217)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:59)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapArray(UntypedObjectDeserializerModule.scala:39)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:233)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:217)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:59)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapArray(UntypedObjectDeserializerModule.scala:39)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:233)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:217)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:59)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapArray(UntypedObjectDeserializerModule.scala:39)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:233)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:217)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:59)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapArray(UntypedObjectDeserializerModule.scala:39)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:233)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:217)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.SeqDeserializer.deserialize(SeqDeserializerModule.scala:59)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapArray(UntypedObjectDeserializerModule.scala:39)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:233)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:495)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3736)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2764)
	at com.fasterxml.jackson.hyperspace.module.scala.ScalaObjectMapper$class.readValue(ScalaObjectMapper.scala:211)
	at com.microsoft.hyperspace.util.JsonUtils$$anon$1.readValue(JsonUtils.scala:30)
	at com.microsoft.hyperspace.util.JsonUtils$.jsonToMap(JsonUtils.scala:43)
	at com.microsoft.hyperspace.index.LogEntry$.fromJson(LogEntry.scala:34)
	at com.microsoft.hyperspace.index.IndexLogManagerImpl.getLog(IndexLogManager.scala:74)
	at com.microsoft.hyperspace.index.IndexLogManagerImpl.getLog(IndexLogManager.scala:78)
	at com.microsoft.hyperspace.index.IndexLogManagerImpl.createLatestStableLog(IndexLogManager.scala:114)
	at com.microsoft.hyperspace.actions.Action$class.end(Action.scala:71)
	at com.microsoft.hyperspace.actions.Action$class.run(Action.scala:94)
	at com.microsoft.hyperspace.actions.CreateAction.run(CreateAction.scala:30)
	at com.microsoft.hyperspace.index.IndexCollectionManager.create(IndexCollectionManager.scala:43)
	at com.microsoft.hyperspace.index.CachingIndexCollectionManager.create(CachingIndexCollectionManager.scala:77)
	at com.microsoft.hyperspace.Hyperspace.createIndex(Hyperspace.scala:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.fasterxml.jackson.core.JsonParseException: Illegal unquoted character ((CTRL-CHAR, code 0)): has to be escaped using backslash to be included in name
 at [Source: java.io.StringReader@51ef8888; line: 980, column: 23]
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1581)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:533)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._throwUnquotedSpace(ParserMinimalBase.java:497)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._parseName2(ReaderBasedJsonParser.java:1464)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._parseName(ReaderBasedJsonParser.java:1431)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextFieldName(ReaderBasedJsonParser.java:778)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:483)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:341)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
	at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
	at com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializer.mapObject(UntypedObjectDeserializerModule.scala:44)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer.deserialize(UntypedObjectDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	... 119 more



```

**Comment by sezruby** [comment]
@mmchougule 

1 => For now, a join query requires:
  - both left and right child (df1, df2) have a candidate index.
  - each candidate index should be built using the same columns of join condition `[id_col, date_col, "another_id"]` as indexed columns.
  - each candidate index should contain all other columns to make the result - so if there's no Project, the index should include all columns in source data.

2 => Did you use any escaped char in file paths or create a partition with a column including that chars?
Seems there is `\0` in a file path; so jackson parser complained about it. 

Could you try with a different source directory path or index system path?

**Comment by sezruby** [comment]
This issue can be resolved by #489. 

@mmchougule Could you try with the fix and let us know if there's still the issue?
Thanks!
